[FEATURE] Add PSR-17 and PSR-18 Documentation (#1037)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/RequestHandling/Index.rst

   Always check the integrity of the middleware stack after changing the default ordering.
   This can be done in the configuration module that comes with EXT:lowlevel.

.. _request-handling-psr-17:

Creating new Request / Response Objects
=======================================

PSR-17_ HTTP Factory interfaces are provided by `psr/http-factory` and should be used as
dependencies for PSR-15_ request handlers or services that need to create PSR-7_ message objects.

It is discouraged to explicitly create PSR-7_ instances of classes from the :php:`\TYPO3\CMS\Core\Http`

> namespace (they are not public API). Use type declarations against PSR-17_ HTTP Message Factory
> interfaces and dependency injection instead.

Reviewer: ```suggestion
namespace (they are not public APIs). Instead, use type declarations against PSR-17_ HTTP Message Factory
interfaces and dependency injections.
```
I could be wrong. Unsure about what is singular and plural here.

Author: "Dependency Injection" should always be singular as it describes the pattern / technology and is kind of a fixed term.

Reviewer: Oh right, cool.

Example
-------

A middleware that needs to send a JSON response when a certain condition is met, uses the

> PSR-17_ response factory interface (the concrete TYPO3 implementation is injected as constructor

Reviewer: ```suggestion
PSR-17_ response factory interface (the concrete TYPO3 implementation is injected as a constructor
```
or 'the' ?

        }

        public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
        {
            if ($request->getRequestTarget() === '/check') {
                $data = ['status' => 'ok'];
                $response = $this->responseFactory->createResponse()
                    ->withHeader('Content-Type', 'application/json; charset=utf-8');
                $response->getBody()->write(json_encode($data));
                return $response;
            }
            return $handler->handle($request);
        }
    }

.. _request-handling-psr-18:

Executing HTTP Requests in Middlewares
======================================

The PSR-18_ HTTP Client is intended to be used by PSR-15_ request handlers in order to perform HTTP
requests based on PSR-7_ message objects without relying on a specific HTTP client implementation.

PSR-18_ consists of a client interface and three exception interfaces:

- :php:`\Psr\Http\Client\ClientInterface`
- :php:`\Psr\Http\Client\ClientExceptionInterface`
- :php:`\Psr\Http\Client\NetworkExceptionInterface`
- :php:`\Psr\Http\Client\RequestExceptionInterface`

> Request handlers shall use dependency injection to retrieve the concrete implementation

Reviewer: ```suggestion
Request handlers will use dependency injection to retrieve the concrete implementation
```
or 'should' ?

Author: how about removing it? `Request handlers use dependency injection ...`?

Reviewer: Yes - nice!

of the PSR-18_ HTTP client interface :php:`\Psr\Http\Client\ClientInterface`.

The PSR-18_ HTTP Client interface is provided by `psr/http-client` and may be used as
dependency for services in order to perform HTTP requests using PSR-7_ request objects.
PSR-7_ request objects can be created with the :ref:`PSR-17 Request Factory interface<request-handling-psr-17>`.

.. note::

>    This does not replace the currently available Guzzle wrapper
>    :php:`\TYPO3\CMS\Core\Http\RequestFactory->request()`, but is available as a framework
>    agnostic, more generic alternative. The PSR-18 interface does not allow to pass request
>    specific guzzle options. But global options defined in :php:`$GLOBALS['TYPO3_CONF_VARS']['HTTP']`
>    are taken into account as GuzzleHTTP is used as backend for this PSR-18 implementation.
>    The concrete implementations is internal and will be replaced by a native guzzle PSR-18
>    implementation once it is available.

Reviewer: ```suggestion
   This does not replace the currently available Guzzle wrapper
   :php:`\TYPO3\CMS\Core\Http\RequestFactory->request()`, but is available as a more generic, 
   framework-agnostic alternative. The PSR-18 interface does not allow you to pass 
   request-specific guzzle options. But global options defined in :php:`$GLOBALS['TYPO3_CONF_VARS']['HTTP']`
   are taken into account because GuzzleHTTP is used as the backend for this PSR-18 implementation.
   The concrete implementation is internal and will be replaced by a native guzzle PSR-18
   implementation once it is available.
```

Example usage
-------------

A middleware might need to request an external service in order to transform the response
into a new response. The PSR-18 HTTP client interface is used to perform the external
HTTP request. The PSR-17 Request Factory Interface is used to create the HTTP request that
the PSR-18 HTTP Client expects. The PSR-7 Response Factory is then used to create a new

> response to be returned to the user. All off these interface implementations are injected

Reviewer: ```suggestion
response to be returned to the user. All of these interface implementations are injected
```
This is a really nicely-worded example 💯
